lib/handlers/pools/provider-migration/v3/traffic-switch-v3-pool-provider.ts

```diff
@@ -0,0 +1,120 @@
+    this.currentPoolProvider = currentPoolProvider
+    this.targetPoolProvider = targetPoolProvider
+    this.sourceOfTruthPoolProvider = sourceOfTruthPoolProvider
+  }
+
+  getPoolAddress(
+    tokenA: Token,
+    tokenB: Token,
+    feeAmount: FeeAmount
+  ): {
+    poolAddress: string
+    token0: Token
+    token1: Token
+  } {
+    // The underlying logic for getting the pool address is always from the
+    // source of truth pool provider, and the result is deterministic.
+    // It doesn't matter which pool provider we use, so we can start with the target
+    // pool provider, which will delegate to the source of truth pool provider.
+    return this.targetPoolProvider.getPoolAddress(tokenA, tokenB, feeAmount)
+  }
+
+  async getPools(tokenPairs: [Token, Token, FeeAmount][], providerConfig?: ProviderConfig): Promise<V3PoolAccessor> {
+    const currentProviderPools = await this.currentPoolProvider.getPools(tokenPairs, providerConfig)
+
+    metric.putMetric('V3_POOL_PROVIDER_POOL_TRAFFIC_TOTAL', 1, MetricLoggerUnit.None)
+    const sampleTraffic = this.SHOULD_SAMPLE_TRAFFIC()
+    if (sampleTraffic) {
+      metric.putMetric('V3_POOL_PROVIDER_POOL_TRAFFIC_SAMPLING', 1, MetricLoggerUnit.None)
+      const targetProviderPools = await this.targetPoolProvider.getPools(tokenPairs, providerConfig)
+
+      // If we need to sample the traffic, we don't want to make it a blocking I/O
+      this.sampleTraffic(tokenPairs, currentProviderPools, targetProviderPools, providerConfig)
+    }
+
+    const switchTraffic = this.SHOULD_SWITCH_TRAFFIC()
+    if (!switchTraffic) {
+      metric.putMetric('V3_POOL_PROVIDER_POOL_TRAFFIC_CURRENT', 1, MetricLoggerUnit.None)
+      return currentProviderPools
+    } else {
+      metric.putMetric('V3_POOL_PROVIDER_POOL_TRAFFIC_TARGET', 1, MetricLoggerUnit.None)
+      return await this.targetPoolProvider.getPools(tokenPairs, providerConfig)
+    }
+  }
+
+  private async sampleTraffic(
+    tokenPairs: [Token, Token, FeeAmount][],
+    currentProviderPools: V3PoolAccessor,
+    targetProviderPools: V3PoolAccessor,
+    providerConfig?: ProviderConfig
+  ): Promise<void> {
+    const truthProviderPools = await this.sourceOfTruthPoolProvider.getPools(tokenPairs, providerConfig)
+
+    truthProviderPools.getAllPools().forEach((pool: Pool) => {
+      const currentProviderPool = currentProviderPools.getPool(pool.token0, pool.token1, pool.fee)
+      if (!currentProviderPool) {
+        // We don't expect missing pool, but export metric in case we see any
+        metric.putMetric('V3_POOL_PROVIDER_POOL_CURRENT_MISSING', 1, MetricLoggerUnit.None)
+        log.info(
+          `v3 Pool ${pool.token0.symbol} ${pool.token1.symbol} ${pool.fee} not found in the current pool provider.`
+        )
+      } else {
+        const sameQuote = currentProviderPool.token0Price.scalar.equalTo(pool.token0Price.scalar)
```

Comment:
we should probably add some jitter here right? exactly the same quote is a tall ask

Reply:
same for the target one

Reply:
I agree that we might need to jitter depending on which pool, based on the cached routes experience from @mikeki . Had a quick chat with @mikeki , and we think we want to ship it to prod to see how the accuracy looks like this week.